clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,59 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = go
+ where
+  go :: HasCallStack => Term -> NormalizeSession Term
+  go e0 = do
+    (e1, hasChanged) <- listen (inlineNonRepWorker e0)
+    if Monoid.getAny hasChanged then
+      pure e1
+    else
+      case e1 of
+        Var _ -> pure e1
+        Data _ -> pure e1
+        Literal _ -> pure e1
+        Prim _ -> pure e1
+        Lam v t -> Lam v <$> go t
+        TyLam v t -> TyLam v <$> go t
+        App a0 b0 -> App <$> go a0 <*> go b0
+        TyApp t typ -> TyApp <$> go t <*> pure typ
+        Cast t0 typ0 typ1 -> Cast <$> go t0 <*> pure typ0 <*> pure typ1
+        Tick tickInfo t -> Tick tickInfo <$> go t
+        Letrec bindings_ body0 -> do
+          let (bndrs, bodies0) = unzip bindings_
+          bodies1 <- mapM go bodies0
+          body1 <- go body0
+          pure (Letrec (zip bndrs bodies1) body1)
+        Case subj0 typ alts -> do
+          (subj1, subjHasChanged) <- listen (go subj0)
+          if Monoid.getAny subjHasChanged then
+            -- This is the crucial part of this transformation, making it
+            -- substantially different from just a bottom-up traversal with
+            -- 'inlineNonRepWorker': if any non representable binder is inlined
+            -- in the the subject of a case expression, we don't traverse the
+            -- alternatives as this inlined structure might very well give use a
+            -- chance to do 'caseCon'. Traversing the alternatives might
+            -- therefore yield a lot of work.
```

Comment:
The comment and the code do not match. The codes says: when the subject of an outer case-statement contains inner case-statements which `inlineNonRep` transforms, then we do not traverse the alternatives of the outer case-statements. But that is not what the comment says.

Reply:
Yeah I'll split this comment out over

```
(e1, hasChanged) <- listen (inlineNonRepWorker e0)
```

and this position to make it clearer.